File: go/mq_batch.go
package homecloud

import (
	"encoding/json"
	"fmt"
)

func entryBodyStr(value any) (string, error) {
	if s, ok := value.(string); ok {
		if s == "" {
			return "", newError("mq.send batch entry body must be non-empty")
		}
		return s, nil
	}
	raw, err := json.Marshal(value)
	if err != nil {
		return "", err
	}
	return string(raw), nil
}

func buildMQBatchEntries(items []any) ([]map[string]any, error) {
	if len(items) < 1 || len(items) > mqBatchMax {
		return nil, newError(fmt.Sprintf("mq.send batch requires 1–%d messages", mqBatchMax))
	}
	entries := make([]map[string]any, 0, len(items))
	seen := map[string]struct{}{}
	for index, item := range items {
		var entry map[string]any
		if m, ok := item.(map[string]any); ok {
			if body, ok := m["body"].(string); ok {
				if body == "" {
					return nil, newError("mq.send batch entry body must be non-empty")
				}
				id := fmt.Sprint(index)
				if m["id"] != nil {
					id = fmt.Sprint(m["id"])
				}
				entry = map[string]any{"id": id, "body": body}
				if h := m["headers"]; h != nil {
					entry["headers"] = h
				}
			}
		}
		if entry == nil {
			body, err := entryBodyStr(item)
			if err != nil {
				return nil, err
			}
			entry = map[string]any{"id": fmt.Sprint(index), "body": body}
		}
		id := fmt.Sprint(entry["id"])
		if _, ok := seen[id]; ok {
			return nil, newError("mq.send batch entry ids must be unique")
		}
		seen[id] = struct{}{}
		entries = append(entries, entry)
	}
	return entries, nil
}

File: java/src/main/java/com/homecloudlab/sdk/MqBatch.java
package com.homecloudlab.sdk;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

final class MqBatch {
    private MqBatch() {}

    static String entryBodyStr(Object value) {
        if (value instanceof String s) {
            if (s.isEmpty()) {
                throw new HomeCloudException("mq.send batch entry body must be non-empty");
            }
            return s;
        }
        try {
            return Json.MAPPER.writeValueAsString(value);
        } catch (Exception e) {
            throw new HomeCloudException("Invalid MQ body", e);
        }
    }

    static List<Map<String, Object>> buildEntries(List<Object> items) {
        if (items == null || items.size() < 1 || items.size() > Env.MQ_BATCH_MAX) {
            throw new HomeCloudException("mq.send batch requires 1–" + Env.MQ_BATCH_MAX + " messages");
        }
        List<Map<String, Object>> entries = new ArrayList<>();
        Set<String> seen = new LinkedHashSet<>();
        for (int index = 0; index < items.size(); index++) {
            Object item = items.get(index);
            Map<String, Object> entry = null;
            if (item instanceof Map<?, ?> m && m.get("body") instanceof String body) {
                if (body.isEmpty()) {
                    throw new HomeCloudException("mq.send batch entry body must be non-empty");
                }
                String id = m.get("id") != null ? String.valueOf(m.get("id")) : Integer.toString(index);
                entry = new LinkedHashMap<>();
                entry.put("id", id);
                entry.put("body", body);
                if (m.get("headers") != null) {
                    entry.put("headers", m.get("headers"));
                }
            }
            if (entry == null) {
                entry = new LinkedHashMap<>();
                entry.put("id", Integer.toString(index));
                entry.put("body", entryBodyStr(item));
            }
            String id = String.valueOf(entry.get("id"));
            if (!seen.add(id)) {
                throw new HomeCloudException("mq.send batch entry ids must be unique");
            }
            entries.add(entry);
        }
        return entries;
    }
}
